benchmarks/benchmarks.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+require "benchmark/ips"
+
+require "memery"
+require "memo_wise"
+require "memoist"
+require "memoized"
+require "memoizer"
+
+BenchmarkGem = Struct.new(:klass, :inheritance_method, :memoization_method)
+
+# We alphabetize this list for easier readability, but shuffle the list before
+# using it to minimize the chance that our benchmarks are affected by ordering.
+BENCHMARK_GEMS = [
+  BenchmarkGem.new(Memery, :include, :memoize),
+  BenchmarkGem.new(MemoWise, :prepend, :memo_wise),
+  BenchmarkGem.new(Memoist, :extend, :memoize),
+  BenchmarkGem.new(Memoized, :include, :memoize),
+  BenchmarkGem.new(Memoizer, :include, :memoize)
+].shuffle
+
+# Use metaprogramming to ensure that each class is created in exactly the
+# the same way.
+BENCHMARK_GEMS.each do |benchmark_gem|
+  Object.const_set("#{benchmark_gem.klass}Example", Class.new do
+    send(benchmark_gem.inheritance_method, benchmark_gem.klass)
+
+    def no_args
+      100
+    end
+    send(benchmark_gem.memoization_method, :no_args)
+
+    def positional_args(a, b)
+      200 + a + b
+    end
+    send(benchmark_gem.memoization_method, :positional_args)
+
+    def keyword_args(a:, b:)
+      300 + a + b
+    end
+    send(benchmark_gem.memoization_method, :keyword_args)
+  end)
+end
+
+# We benchmark different cases separately, to ensure that slow performance in
```

Comment:
Is there a better way to group these benchmarks?

Reply:
This would be more metaprogramming, but I wonder if it'd be more complex or more straightforward to store a map of proc name --> block to execute and then iterate through that map calling the benchmark functionality over it?

Reply:
Talked over Zoom; we couldn't find a good way to do this.